lib/howl/interactions/replacement.moon

```diff
@@ -0,0 +1,390 @@
+
+    @_clear_highlights!
+
+    return unless @num_matches > 0
+
+    -- skip until visible section
+    local visible_start
+    for i = 1, @num_matches
+      preview_position = @adjusted_positions[i] or match.start_pos
+      if preview_position >= first_visible
+        visible_start = i
+        break
+
+    -- highlight visible section only
+    return unless visible_start
+    for i = visible_start, @num_matches
+      match = @matches[i]
+      preview_position = @adjusted_positions[i] or match.start_pos
+      if preview_position > last_visible
+        break
+      hlt = @selected_idx == i and 'search' or 'search_secondary'
+      len = @replacements_applied[i] and @replacements_applied[i].ulen or match.end_pos - match.start_pos + 1
+      @_highlight_match hlt, preview_position, len
+      if @strikeouts[i]
+        @_highlight_match 'replace_strikeout', preview_position, len
+
+  _clear_highlights: (start_idx=1) =>
+    highlight.remove_all 'search', @preview_buffer
+    highlight.remove_all 'search_secondary', @preview_buffer
+    highlight.remove_all 'replace_strikeout', @preview_buffer
+
+  _highlight_match: (name, pos, len) =>
+    highlight.apply name, @preview_buffer, pos, len
+
+  _toggle_current: =>
+    newval = not @excluded[@selected_idx]
+    @excluded[@selected_idx] = newval
+    @num_excluded += newval and 1 or -1
+    @_preview_replacements @selected_idx
+    @_update_caption!
+
+  _switch_to: (cmd) =>
+    captured_text = @command_line.text
+    @command_line\run_after_finish ->
+      app.editor.selection\select @start_pos, @end_pos
+      command.run cmd .. ' ' .. captured_text
+    @_restore_return!
+
+  keymap:
+    escape: => @_restore_return!
+
+    alt_enter: => @_toggle_current!
+
+    enter: => @runner\run 'submit', ->
+      cursor_pos = @selected_idx and @matches[@selected_idx] and @matches[@selected_idx].start_pos
+      result =
+        num_replaced: @num_replacements - @num_excluded
+        target: @target
+        replacement: @replacement
+        line_at_top: app.editor.buffer.line_at_top
+        :cursor_pos
+
+      if result.num_replaced > 0
+        @_preview_replacements 1, false
+        result.text = @preview_buffer.text
+
+      @_restore_return result
+
+    ctrl_r: => @_switch_to 'buffer-replace-regex'
+
+    binding_for:
+      ["cursor-down"]: =>
+        return unless (@num_matches > 0) and @selected_idx
+        if @selected_idx == @num_matches
+          @selected_idx = 1
+        else
+          @selected_idx += 1
+        @_preview_highlights!
+
+      ["cursor-up"]: =>
+        return unless (@num_matches > 0) and @selected_idx
+        if @selected_idx == 1
+          @selected_idx = @num_matches
+        else
+          @selected_idx -= 1
+        @_preview_highlights!
+
+      ['editor-scroll-up']: -> app.editor\scroll_up!
+      ['editor-scroll-down']: -> app.editor\scroll_down!
+
+      ['buffer-replace']: => @_switch_to 'buffer-replace'
```

Comment:
As a general note, the ability to switch between the different replacement modes is nice, but we should extend it in the future. As an example, it would be really useful to be able to switch to replacement from any of the search commands, e.g. `buffer-search-forward`, then the key stroke for replacement to enter into replacement mode scoped to the previous matches.